script/mafia

```diff
@@ -0,0 +1,75 @@
+#!/bin/sh -eu
+
+fetch_latest () {
+  if [ -z ${MAFIA_TEST_MODE+x} ]; then
+    TZ=$(date +"%T")
+    curl --silent "https://raw.githubusercontent.com/ambiata/mafia/master/script/mafia?$TZ"
+  else
+    cat ../script/mafia
+  fi
+}
+
+run_upgrade () {
+  echo "Checking for a new version of mafia ..."
+  fetch_latest > /tmp/mafia
+
+  COMMIT_VERSION=$(git ls-remote https://github.com/ambiata/mafia | grep refs/heads/master | cut -f 1)
+  echo "# Version: $COMMIT_VERSION" >> /tmp/mafia
+
+  if ! cmp ./mafia /tmp/mafia >/dev/null 2>&1; then
+    echo "New version found and upgraded. You can now commit it to your git repo."
+    mv /tmp/mafia ./mafia
+    chmod +x ./mafia
+  else
+    echo "You have latest mafia script"
+  fi
+}
+
+exec_mafia () {
+  MAFIA_VERSION=$(awk '/^# Version: / { print $3; exit 0; }' $0)
+  MAFIA_BASE=$HOME/.ambiata/mafia/bin/mafia
+  MAFIA_PATH=$MAFIA_BASE-$MAFIA_VERSION
+
+  [ -f "$MAFIA_PATH" ] || {
+    MAFIA_TEMP=$(mktemp -d 2>/dev/null || mktemp -d -t 'exec_mafia')
+    trap "rm -rf \"$MAFIA_TEMP\"" EXIT
+
+    ( cd "$MAFIA_TEMP"
+
+      git clone https://github.com/ambiata/mafia
+      cd mafia
+
+      git reset --hard $MAFIA_VERSION
+
+      SUBMODULES=$(find lib -mindepth 1 -maxdepth 1)
+      for SUBMODULE in $SUBMODULES; do
+        git submodule update --init -- $SUBMODULE
+      done
+
+      cabal sandbox init
+
+      CABAL_SOURCES=$(find lib -maxdepth 4 ! -path lib/\*/bin/\* ! -path lib/\*/lib/\* -name \*.cabal | xargs -L 1 dirname)
+      for CABAL_SOURCE in $CABAL_SOURCES; do
+        cabal sandbox add-source -- $CABAL_SOURCE
+      done
+
+      cabal install )
+
+    mkdir -p $(dirname $MAFIA_PATH)
+    cp "$MAFIA_TEMP/mafia/.cabal-sandbox/bin/mafia" "$MAFIA_PATH"
+  }
```

Comment:
Not sure, but I think a lot of this might be easier to maintain outside of this script, as in, if this script just did a checkout, and than ran ./bootstrap or similar - would make it easier to change etc.... Not a big deal though.